File: concepts/08_async_and_concurrency/04_threads_workers_and_process_isolation/cpp/test_02_result_and_transfer_boundaries.cpp
// 并发结果、异常与数据传输边界。
// 共同问题：工作单元失败如何回到调用方；普通对象是共享、复制还是序列化；
// 所有者如何确认工作单元已经结束。
//
// polyglot-family: async_and_concurrency
// polyglot-concept: threads_workers_and_process_isolation
// polyglot-related: languages/cpp/standard_library/17_concurrency/
// polyglot-related+: test_147_thread_creation_arguments_ids_join_detach_and_this_thread.cpp
// polyglot-related: languages/cpp/standard_library/17_concurrency/
// polyglot-related+: test_154_promise_future_shared_state_results_errors_and_waiting.cpp

#include <gtest/gtest.h>

#include <future>
#include <stdexcept>
#include <thread>

namespace {

TEST(ThreadResultConcept, AsyncFutureRethrowsWorkerExceptionToOwner) {
  auto result = std::async(std::launch::async, []() -> int {
    throw std::runtime_error{"worker failed"};
  });

  EXPECT_THROW(static_cast<void>(result.get()), std::runtime_error);
}

TEST(ThreadResultConcept, ThreadCanObserveTheSameObjectAddress) {
  int value = 1;
  const int* observed_address = nullptr;
  std::thread worker{[&] {
    observed_address = &value;
    value = 42;
  }};

  worker.join();

  EXPECT_EQ(observed_address, &value);
  EXPECT_EQ(value, 42);

  // std::thread 共享地址空间；join 建立完成同步。C++20 标准库没有进程/Worker 与
  // 结构化克隆边界，跨进程创建和 IPC 需要平台或第三方 API。
}

}  // namespace

File: concepts/08_async_and_concurrency/04_threads_workers_and_process_isolation/python/test_02_result_and_transfer_boundaries.py
"""并发结果、异常与数据传输边界。

共同问题：工作单元失败如何回到调用方；普通对象是共享、复制还是序列化；
所有者如何确认工作单元已经结束。
"""

# polyglot-family: async_and_concurrency
# polyglot-concept: threads_workers_and_process_isolation
# polyglot-related: languages/python/stdlib/088-094_concurrent_execution/
# polyglot-related+: test_092_concurrent_futures_complete_states_waiting_thread_and_process_pools.py

from concurrent.futures import ThreadPoolExecutor
import multiprocessing

import pytest


def _mutate_and_send(connection, payload):
    payload.append(3)
    connection.send(payload)
    connection.close()


def test_thread_pool_future_rethrows_worker_exception_to_owner():
    failure = RuntimeError("worker failed")

    def fail():
        raise failure

    with ThreadPoolExecutor(max_workers=1) as executor:
        result = executor.submit(fail)

        with pytest.raises(RuntimeError) as caught:
            result.result()

    assert caught.value is failure


def test_process_pipe_serializes_value_instead_of_sharing_list_identity():
    context = multiprocessing.get_context("fork")
    original = [1, 2]
    receiver, sender = context.Pipe(duplex=False)
    worker = context.Process(target=_mutate_and_send, args=(sender, original))

    worker.start()
    sender.close()
    transferred = receiver.recv()
    receiver.close()
    worker.join()

    assert worker.exitcode == 0
    assert transferred == [1, 2, 3]
    assert original == [1, 2]

    # fork 初始内存快照与 Pipe pickle 都不共享 list identity；需要共享状态时必须选用
    # multiprocessing.Value/Array/shared_memory 并定义同步。
